algorithms/nautilus/grammartec/context.cpp

```diff
@@ -0,0 +1,495 @@
+ */
+void Context::CalcNumOptions() {
+  for (auto& elem: _nts_to_rules) {
+    if (_nts_to_num_options.find(elem.first) == _nts_to_num_options.end()) {
+      _nts_to_num_options[elem.first] = _rules.size();
+    }
+  }
+
+  bool something_changed;
+  do {
+    something_changed = false;
+
+    for (size_t i = 0; i < _rules.size(); i++) {
+      RuleID rid(i);
+      size_t num = CalcNumOptionsForRule(rid);
+      const NTermID& nt = GetRule(rid).Nonterm();
+
+      if (_nts_to_num_options.find(nt) == _nts_to_num_options.end())
+        _nts_to_num_options[nt] = num;
+
+      /* Update maximum number */
+      if (_nts_to_num_options[nt] < num) {
+        _nts_to_num_options[nt] = num;
+        something_changed = true;
+      }
+
+      _rules_to_num_options[rid] = num;
+    }
+  } while (something_changed);
+}
+
+/**
+ * @fn
+ * Calculate the minimum length for a rule
+ * @brief Calculate minimum length
+ * @param (r) RuleID
+ * @return Minimum length (nullopt on failure)
+ */
+std::optional<size_t> Context::CalcMinLenForRule(const RuleID& r) const {
+  size_t res = 1;
+
+  for (const NTermID& nt_id: GetRule(r).Nonterms()) {
+    if (_nts_to_min_size.find(nt_id) == _nts_to_min_size.end()) {
+      return std::nullopt;
+
+    } else {
+      res += _nts_to_min_size.at(nt_id);
+    }
+  }
+
+  return res;
+}
+
+/**
+ * @fn
+ * @brief Sort rules
+ */
+void Context::CalcRuleOrder() {
+  for (auto& elem: _nts_to_rules) {
+    std::vector<RuleID>& rules = elem.second;
+    std::sort(rules.begin(), rules.end(),
+              [this](RuleID& r1, RuleID& r2) {
+                return _rules_to_min_size.at(r1) < _rules_to_min_size.at(r2);
+              });
+  }
+}
+
+/**
+ * @fn
+ * @brief Calculate minimum length
+ * @throw exceptions::fuzzuf_runtime_error Grammar is invalid
+ */
+void Context::CalcMinLen() {
+  bool something_changed;
+
+  do {
+    std::vector<RuleID> unknown_rules;
+    unknown_rules.reserve(_rules.size());
+    for (size_t i = 0; i < _rules.size(); i++) {
+      unknown_rules.emplace_back(i);
+    }
+    something_changed = false;
+
+    while (unknown_rules.size()) {
+      size_t last_len = unknown_rules.size();
+
+      /* Remove every rule with known minimum length */
+      for (auto it = unknown_rules.crbegin();
+           it != unknown_rules.crend();
+           ++it) {
+        const RuleID& rule = *it;
+
+        if (std::optional<size_t> min = CalcMinLenForRule(rule)) {
+          const NTermID& nt = GetRule(rule).Nonterm();
+
+          if (_nts_to_min_size.find(nt) == _nts_to_min_size.end())
+            _nts_to_min_size[nt] = min.value();
+
+          if (_nts_to_min_size[nt] > min.value()) {
+            /* Update minimum value */
+            _nts_to_min_size[nt] = min.value();
+            something_changed = true;
+          }
+
+          _rules_to_min_size[rule] = min.value();
+
+          /* Be careful base of reverse iterator is off by one */
+          unknown_rules.erase(--(it.base()));
```

Comment:
As far as I understand, this is invalid in the specifications. You can receive the return value(which is an iterator) instead.
Moreover, probably you can write this procedure a bit similarly to Rust by using `vector::remove` with `std::remove_if`.

Reply:
Fixed in [80717f4](https://github.com/fuzzuf/fuzzuf/pull/23/commits/80717f42d2b634a8c9d5b242aca7a897c2a78305)